gems/aws-sdk-s3/lib/aws-sdk-s3/multipart_stream_uploader.rb

```diff
@@ -62,12 +62,18 @@ def upload_parts(upload_id, options, &block)
         completed = Queue.new
         errors = IO.pipe do |read_pipe, write_pipe|
           threads = upload_in_threads(read_pipe, completed, upload_part_opts(options).merge(upload_id: upload_id))
-          block.call(write_pipe)
-          write_pipe.close
+          begin
+            block.call(write_pipe)
+          ensure
+            write_pipe.close # Ensure the pipe is closed to avoid https://github.com/jruby/jruby/issues/6111
+          end
           threads.map(&:value).compact
         end
+      rescue => e
+        errors = [e]
+      ensure
         if errors.empty?
-          Array.new(completed.size) { completed.pop }.sort_by { |part| part[:part_number] }
+          return Array.new(completed.size) { completed.pop }.sort_by { |part| part[:part_number] }
```

Comment:
Minor: why the explicit return?

Reply:
Explicit return [is significant](https://blog.leshill.org/blog/2009/11/17/ensure-with-explicit-return.html) in `ensure` clauses and is needed to properly return the parts list. I opted for this variant to avoid nesting a `begin` inside another `begin`.

This violates the Principle of least astonishment and is also [discouraged by rubocop](https://github.com/rubocop-hq/ruby-style-guide#no-return-ensure), so I've changed the code to avoid the `ensure` entirely.

Reply:
Ah interesting! Thanks for sharing, I did not know the differences here.